core/domain/exp_domain.py

```diff
@@ -1343,8 +1348,7 @@ def validate(self, exp_param_specs_dict):
 
         if self.hints:
             if self.solution:
-                Solution.from_dict(
-                    self.id, self.solution).validate(self.id)
+                self.solution.validate(self.id)
```

Comment:
Regardless of the link between hints and solution, might as well just do these validations separately anyway? There can be a separate explicit check for the conditions under which each is allowed to exist or not.